### PR TITLE
Pytorch operator proposal

### PR DESCRIPTION
**Pull request by jose5918**
Currently a work in progress proposal for pytorch-operator. I ran distributed pytorch on kubernetes with similar specs so I wanted to start the discussion.

I intend have a POC by the end of next week

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/community/33)

### CODE REVIEW

File: proposals/pytorch-operator-proposal.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+apiVersion: "kubeflow.org/v1alpha1"
+kind: "PytorchJob"
+metadata:
+  name: "example-job"
+spec:
+  backend: "gloo"
+  masterPort: "23456"
+  replicaSpecs:
+    - replicas: 1
+      ReplicaType: MASTER
```
Comment (ScorpioCPH) [inline proposals/pytorch-operator-proposal.md]: FYI, tf-operator will use this struct:
```go
map[ReplicaType]*ReplicaSpec
```

File: proposals/pytorch-operator-proposal.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
+## Motivation
+Pytorch is a popular machine learning framework which currently does not have an operator/controller for Kubernetes. This proposal is aimed at defining what that operator should look like, and adding it to Kubeflow.
```
Comment (jlewi) [inline proposals/pytorch-operator-proposal.md]: My biggest question about pytorch is how to do distributed communication. Based on my limited research 
kubeflow/kubeflow#179 it looks like there are multiple backends e.g.Gloo and MPI.

**Comment by jlewi** [comment]
I think this is a great idea.

I'd love to see work supporting PyTorch start as soon as possible; whether distributed or not (e.g. just adding it to our notebook images). I'll leave that to whoever wants to work on this.

So I don't think we should block development waiting for a fully fleshed out proposal. I'm fine submitting the proposal as is or leaving it as an open PR which ever people think best.

However, @jose5918 if/when you're ready to start coding I suggest we just create a repo "experimental-pytorch" and let people who are interested start working.

File: proposals/pytorch-operator-proposal.md (hunk 2 above)
Comment (Jimexist) [inline proposals/pytorch-operator-proposal.md]: should we use `PyTorch` instea of `Pytorch`?

File: proposals/pytorch-operator-proposal.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+- A distributed pytorch example 
+
+## Non-Goals
+For the scope of this proposal, we won't be addressing the method for serving the model.
+
+## API (CRD and resulting objects)
+
+### Custom Resource Definition
+The custom resource submitted to the Kubernetes API would look something like this:
+```
```
Comment (Jimexist) [inline proposals/pytorch-operator-proposal.md]: add `yaml` after three ticks

File: proposals/pytorch-operator-proposal.md (hunk 2 above)
Comment (jose5918) [reply proposals/pytorch-operator-proposal.md]: Yeah I'll update

Comment (jose5918) [reply proposals/pytorch-operator-proposal.md]: I've tried it with TCP and Gloo in Kubernetes and it seems to work okay, but I need to build PyTorch with MPI already installed if I want to try it. I also don't have a GPU cluster so I'm probably not seeing the bigger benefits of the different backends. I was thinking just provide the backend as an environment variable and people can choose to use that environment variable in their code if they want or just hard code their backend.

Comment (jose5918) [reply proposals/pytorch-operator-proposal.md]: I'm gonna have to take back the gloo comment. It's definitely not working right aside from connecting

**Comment by jose5918** [comment]
@jlewi I have started something already, but what would be the correct process to create a repo? Creating an issue like this? https://github.com/kubeflow/community/issues/35

File: proposals/pytorch-operator-proposal.md (hunk 2 above)
Comment (jlewi) [reply proposals/pytorch-operator-proposal.md]: @jose5918 If you need GPUs just let me know you can use dev.kubeflow.org.

File: proposals/pytorch-operator-proposal.md (hunk 1 above)
Comment (jose5918) [reply proposals/pytorch-operator-proposal.md]: Yeah I think I will change to use a map

**Comment by jlewi** [comment]
/lgtm

My suggestion is we accept the proposal and iterate. Does someone else want to provide a second lgtm?

**Comment by soumith** [comment]
hey guys, thanks a lot for the proposal. I have reviewed it and found no obvious holes wrt pytorch perspective.

**Pull request merged**